aip/general/0193.md

```diff
@@ -458,6 +327,85 @@ If the user does have proper permission, but the requested resource or
 parent does not exist, the service **must** error with `NOT_FOUND` (HTTP
 404).
 
+## HTTP/1.1+JSON representation
+
+When clients use HTTP/1.1 as per [AIP-127](./0127.md), the error information
+is returned in the body of the response, as a JSON object. For backward
+compatibility reasons, this does not map precisely to `google.rpc.Status`,
+but contains the same core information. The schema is defined in the following proto:
+
+```proto
+message Error {
+  message Status {
+    // The HTTP status code that corresponds to `google.rpc.Status.code`.
+    int32 code = 1;
+    // This corresponds to `google.rpc.Status.message`.
+    string message = 2;
+    // This is the enum version for `google.rpc.Status.code`.
+    google.rpc.Code status = 4;
+    // This corresponds to `google.rpc.Status.details`.
+    repeated google.protobuf.Any details = 5;
+  }
+
+  Status error = 1;
+}
+```
+
+The most important difference is that the `code` field in the JSON is an HTTP status code,
+*not* the direct value of `google.rpc.Status.code`. For example, a `google.rpc.Status`
+message with a `code` value of 5 would be mapped to an object including the following
+code-related fields (as well as the message, details etc):
+
+```json
+{
+  "error": {
+    "code": 404,          // The HTTP status code for "not found"
+    "status": "NOT_FOUND" // The name in google.rpc.Code for value 5
+  }
+}
+```
+
+The following JSON shows a fully populated HTTP/1.1+JSON representation of an error response.
+
+
+```json
+{
+  "error": {
+    "code": 429,
+    "message": "The zone 'us-east1-a' does not have enough resources available to fulfill the request. Try a different zone, or try again later.",
+    "status": "RESOURCE_EXHAUSTED",
+    "details": [
+      {
+        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
+        "reason": "RESOURCE_AVAILABILITY",
+        "domain": "compute.googleapis.com",
+        "metadata": {
+          "zone": "us-east1-a",
+          "vmType": "e2-medium",
+          "attachment": "local-ssd=3,nvidia-t4=2",
+          "zonesWithCapacity": "us-central1-f,us-central1-c"
+        }
+      },
```

Comment:
Is the `ErrorInfo` optional? I wonder why it wouldn't be part of the `error` itself instead of `details`

Likewise with other things like localized message, if they suppose to be only one entry, since `details` is a list of anything so, it could have many error info or localized messages.

Sorry for the confusion.

Reply:
As noted earlier in the AIP:

- There must *always* be an ErrorInfo. (I suspect that changing `Status` itself to include that directly would be hard in various ways.)
- There must only ever be one instance of any given message type. Yes, that's not a constraint that's expressed in the schema itself (because it can't be) but is specified in the text.

Reply:
> Yes, that's not a constraint that's expressed in the schema itself (because it can't be) but is specified in the text.

Wouldn't it help the engineers to make such an invariant part of the schema itself? 

There are some breaking changes, so it is not that easy to adopt or change, but I am just curious about the motivation for using `details` for it overall.

Reply:
> Wouldn't it help the engineers to make such an invariant part of the schema itself?

There's no way of saying "This can use arbitrary messages, but only once per message." We could have had a message with all the "known" additional info messages, but that would have been limiting in terms of APIs wanting to add their own information outside these AIPs.

Yes, in an alternative universe we could have made `ErrorInfo` a field in `Status` - but that's not where we are, and that choice is not up for debate.

Reply:
In a vacuum, @yordis, yes adding a concrete field to the schema would benefit developers, but there are a few points to using `details`...

1. The reality is that `google.rpc.Status` is used in a number of non-Google-specific ways, so shipping further Google-isms in the schema may be undesirable
2. `details` gives us the flexibility to add new properties without versioning `google.rpc.Status` (which would be infeasible given the breadth of usage)
3. `details` gives *anyone* i.e. non-Google users, the flexibility to add new properties without needing to change `google.rpc.Status`

Without a governance model around `google.rpc.Status` similar to that of HTTP protocol, further extending `google.rpc.Status` concretely would be a difficult task. Instead, we invest in our 1p clients offered to customers to present these well-known details in a reasonable way.

Reply:
@noahdietz @jskeet, I sincerely appreciate your taking the time to explain a bit more. 

> Yes, in an alternative universe we could have made ErrorInfo a field in Status - but that's not where we are, and that choice is not up for debate.

That is precisely the confirmation I needed. It has been almost half a decade since I reverse-engineered and understood every significant business's specifications, and I took massive inspiration from AIP-193 for https://github.com/straw-hat-team/adr/pull/30 (if you have some spare time, I would deeply appreciate your participation on it, just a second try in that alternative universe).

I just wasn't confident enough about this particular scenario. Seriously, thank you so much, years trying to figure out the WHY behind it!

Reply:
For sure, good luck @yordis, cheers!